server/controllers/calls.js

```diff
@@ -194,17 +229,25 @@ export function recordAttempt(req, res) {
               return putCallAttempt(call_id, outcome, notes)
                 .then(() => {
                   const attempt_num = parseInt(call.attributes.attempt_num, 10);
-                  if (responses) {
-                    Promise.all(responses.map((resp) => {
-                      const { question_id, response } = resp;
-                      const responseParams = { call_id, question_id, response };
-                      return responsesService.saveNewResponse(responseParams);
-                    }))
-                    .then(() => afterPut(res, outcome, contact_id, attempt_num, campaign_id))
-                    .catch(err => console.log('error saving responses in recordAttempt function of calls controller when saving a call attempt :', err));
-                  }
-                  return afterPut(res, outcome, contact_id, attempt_num, campaign_id)
-                    .then().catch(err => console.log('error creating new call in log for required subsequent contact after recording attempt in recordAttempt functino of calls controller: ', err));
+                  return isLastAttemptedCall(campaign_id, outcome)
+                    .then((isLastCall) => {
+                      if (!isLastCall) {
+                        if (responses) {
+                          return saveResponses(responses, call_id)
+                            .then(() => res.status(201).json({ message: 'Successfully saved new responses' }))
+                            .catch(err => res.status(500).json({ message: `Could not save new responses: ${err}` }));
+                        }
+                        return afterPut(res, outcome, contact_id, attempt_num, campaign_id)
+                          .then(() => res.status(200).json({ message: 'Contact and call log updated successfully.' }))
+                          .catch(err => console.log('Could not update contact status or call log: ', err));
+                      }
+                      if (responses) {
+                        saveResponses(responses, call_id);
```

Comment:
Not sure that I follow what is happening here. Do we intend for the conditional to run and then the return statement at the bottom to run?
If there is logic that should happen after this function runs, why is it not in a .then block?

Reply:
Yes, that was the plan! Is it better to return here and then do two .thens (one after saveResponses and one after afterPut) with the markCampaignComplete?

Reply:
@andipants12 made this update! can you take a look and let me know what you think. I think you're right that it makes it a bit more clear as to what's going on.

Reply:
@allieglotfelty Looks good 😸